app/Main.hs

```diff
@@ -0,0 +1,151 @@
+import Text.Read (readMaybe)
+
+type Coordinates = (Int, Int)
+
+data Player = Cross | Circle deriving (Eq, Show)
+
+showPlayerT :: Player -> Text.Text
+showPlayerT = Text.pack . show
+
+newtype Board = Board (Map Coordinates Player)
+
+showBoard :: Board -> Text.Text
+showBoard (Board board) = foldl' replaceExclamation unfilledBoard $ flip (findWithDefault " ") textBoard <$> coordinates
+  where
+    playerToSpotText :: Player -> Text.Text
+    playerToSpotText Circle = "0"
+    playerToSpotText Cross = "x"
+
+    textBoard = playerToSpotText <$> board
+    -- Not Sure what I think about this implementations but it works and is pretty clean
+    coordinates :: [Coordinates]
+    coordinates = [(x, y) | y <- [0 .. 2], x <- [0 .. 2]]
+
+    replaceExclamation :: Text.Text -> Text.Text -> Text.Text
+    replaceExclamation boardText spotText = replaceOne "!" spotText boardText
+
+    unfilledBoard =
+      "  0   1   2 \n\
+      \0 ! | ! | ! \n\
+      \ ---+---+---\n\
+      \1 ! | ! | ! \n\
+      \ ---+---+---\n\
+      \2 ! | ! | ! "
+
+nextTurn :: Player -> Player
+nextTurn Cross = Circle
+nextTurn Circle = Cross
+
+replaceOne :: Text.Text -> Text.Text -> Text.Text -> Text.Text
+replaceOne pattern substitution text = Text.concat [front, substitution, Text.drop (Text.length pattern) back]
+  where
+    (front, back) = Text.breakOn pattern text
+
+playAtSpot :: Coordinates -> Player -> Board -> Maybe Board
+playAtSpot coord spot (Board board) = case board !? coord of
+  Nothing -> Just $ Board $ insert coord spot board
+  Just _ -> Nothing
+
+unrepeat :: Eq a => [a] -> Maybe a
+unrepeat (x : xs) = guard (all (== x) xs) >> pure x
+unrepeat _ = Nothing
+
+winner :: Board -> Coordinates -> Maybe Player
+winner (Board board) (x, y) =
+  foldl1' (<|>) $ isWin <$> winningChecks
+  where
+    winningChecks
+      -- Center
+      | x == 1 && y == 1 = [row, column, diag2, diag1]
+      -- True if coordinates aren't part of diagonal
+      | (x == 1 || y == 1) && (x /= y) = [row, column]
+      -- True for diag 1
+      | x == y = [row, column, diag1]
+      -- True for diag 2
+      | otherwise = [row, column, diag2]
+
+    isWin = unrepeat <=< traverse (board !?)
+
+    diag1 = [(n, n) | n <- [0 .. 2]]
+    diag2 = [(2 - n, n) | n <- [0 .. 2]]
+    row = [(x', y) | x' <- [0 .. 2]]
+    column = [(x, y') | y' <- [0 .. 2]]
+
+getCoordsAndPlayTurn :: Board -> Player -> IO (Board, Coordinates)
+getCoordsAndPlayTurn board turn = do
+  TextIO.putStrLn $ showPlayerT turn <> " turn"
+  coord <- getInputAndValidate parseIntTuple "Please Enter space seperated coordinates: x y"
+  case playAtSpot coord turn board of
+    Nothing -> do
+      setSGR [SetColor Foreground Dull Red]
+      TextIO.putStrLn "Invalid placement"
+      setSGR [Reset]
+      getCoordsAndPlayTurn board turn
+    Just newBoard -> pure (newBoard, coord)
+
+type Parser a = (Text.Text -> Either Text.Text a)
```

Comment:
If you wanna play around with parsers, have a look at [`attoparsec`](https://hackage.haskell.org/package/attoparsec)